### ISSUE TITLE
async fn in trait: ICE with 2 generic params in method

### ISSUE THREAD

**Issue by Dirbaio** [rust-lang/rust#102310]
[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=e125577e1190890f2ddbb3213c4fb6cb)
```rust
pub trait SpiDevice {
    async fn transaction<F, R>(&mut self);
}

impl SpiDevice for () {
    async fn transaction<F, R>(&mut self) {}
}
```

gives 

```
error: internal compiler error: compiler/rustc_middle/src/ty/subst.rs:674:9: expected type for `R/#1` (R/1) but found Lifetime(ReFree(DefId(0:11 ~ embedded_hal_async[dae4]::spi::{impl#0}::transaction), BrNamed(DefId(0:16 ~ embedded_hal_async[dae4]::spi::SpiDevice::transaction::'_), '_))) when substituting, substs=[R, ReFree(DefId(0:11 ~ embedded_hal_async[dae4]::spi::{impl#0}::transaction), BrNamed(DefId(0:16 ~ embedded_hal_async[dae4]::spi::SpiDevice::transaction::'_), '_)), ReFree(DefId(0:11 ~ embedded_hal_async[dae4]::spi::{impl#0}::transaction), BrNamed(DefId(0:16 ~ embedded_hal_async[dae4]::spi::SpiDevice::transaction::'_), '_))]
```

[full stack trace](https://gist.github.com/Dirbaio/fac2b7995a7f9bdd5cbc0d88f99c06ce)

happens with either

rustc 1.66.0-dev (21265dd0d2)
1.66.0-nightly (2022-09-25 f5193a9fcc73dc09e41a)

**Comment by Rageking8** [comment]
@rustbot label +T-compiler +C-bug +I-ICE +requires-nightly

@Dirbaio Please use the ICE issue template for subsequent ICE issues. Thanks.

**Comment by compiler-errors** [comment]
This is fixed by #102334
